gk/main.c

```diff
@@ -16,22 +16,404 @@
+			 * expires.
+			 */
+			uint64_t expire_at;
+		} declined;
+	} u;
+};
+
+static uint8_t
+integer_log_base_2(uint64_t delta_time)
+{
+	return (8 * sizeof(uint64_t) - __builtin_clzll((delta_time)) - 1);
+}
+
+/* 
+ * It converts the difference of time between the current packet and 
+ * the last seen packet into a given priority. 
+ */
+static uint8_t 
+priority_from_delta_time(uint64_t present, uint64_t past)
+{
+	uint64_t delta_time;
+
+	if (unlikely(present < past)) {
+		/*
+		 * This should never happen, but we handle it gracefully here 
+		 * in order to keep going.
+		 */
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: the present time smaller than the past time!\n");
+
+		return 0;
+	}
+
+	delta_time = (present - past) * picosec_per_cycle;
+	if (unlikely(delta_time < 1))
+		return 0;
+	
+	return integer_log_base_2(delta_time);
+}
+
+static inline void
+extract_packet_info(struct rte_mbuf *pkt, struct ipacket *packet)
+{
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr  *ip4_hdr;
+	struct ipv6_hdr  *ip6_hdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+
+		ip4_hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv4_hdr *, 
+					sizeof(struct ether_hdr));
+
+		packet->flow.proto = ETHER_TYPE_IPv4;
+		packet->flow.f.v4.src = rte_be_to_cpu_32(ip4_hdr->src_addr);
+		packet->flow.f.v4.dst = rte_be_to_cpu_32(ip4_hdr->dst_addr);
+
+		break;
+
+	case ETHER_TYPE_IPv6:
+
+		ip6_hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv6_hdr *, 
+					sizeof(struct ether_hdr));
+
+		packet->flow.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(packet->flow.f.v6.src, ip6_hdr->src_addr, 16);
+		rte_memcpy(packet->flow.f.v6.dst, ip6_hdr->dst_addr, 16);
+
+		break;
+
+	default:
+
+		packet->flow.proto = 0;
+
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"gk: unknown network layer protocol!\n");
+	}
+}
+
+static inline void
+initialize_flow_entry(struct flow_entry *fe, struct ip_flow *flow)
+{
+	rte_memcpy(&fe->flow, flow, sizeof(*flow));
+
+	fe->state = GK_REQUEST;
+	fe->u.request.last_packet_seen_at = rte_rdtsc();
+	fe->u.request.last_priority = START_PRIORITY;
+	fe->u.request.allowance = START_ALLOWANCE - 1;
+	/* TODO Grantor ID comes from LPM lookup. */
+	fe->u.request.grantor_id = 0;
+}
+
+static int
+drop_packet(struct rte_mbuf *pkt)
+{
+	rte_pktmbuf_free(pkt);
+	return 0;
+}
+
+/* 
+ * When a flow entry is at request state, all the GK block processing
+ * that entry does is to:
+ * (1) compute the priority of the packet.
+ * (2) encapsulate the packet as a request.
+ * (3) put this encapsulated packet in the request queue.
+ */
+static int
+gk_process_request(struct flow_entry *fe, struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+	uint8_t priority = priority_from_delta_time(now,
+			fe->u.request.last_packet_seen_at);
+
+	fe->u.request.last_packet_seen_at = now;
+	if (priority <= fe->u.request.last_priority &&
```

Comment:
Notice that we are not talking about transferring packets here, but only discussing the association of priorities. If the priority of a packet is enough to not be dropped in the priority queue, it will arrive at the grantor server. If the interval between two packets leads to a priority that is equal to the last priority assigned, it means that the source has waited enough time to be granted that priority, so it should also be entitled to reset its allowance.

If you have a concrete case in mind that shows it's not the proper choice, please describe it precisely, so we can go over and see what happens and decide if it's a good or bad outcome.